### PR TITLE
Integration tests

### PR DESCRIPTION
**Pull request by levackt**
Still a couple of tests failing with Unreachable, eg;

```
---- happy_days_end_poll stdout ----
thread 'happy_days_end_poll' panicked at 'VM error: 
RuntimeErr { msg: "Wasmer runtime error: InvokeError(TrapCode { code: 

Unreachable, srcloc: 208452 })", backtrace: Backtrace(()) }', 
/Users/taariq/.cargo/registry/src/github.com-1ecc6299db9ec823/cosmwasm-vm-0.9.2/src/testing/calls.rs:43:5
```

### CODE REVIEW

**Comment by ethanfrey** [comment]
Wow, I have never seen that error before, but I did reproduce locally. Investigating now...

```
thread 'end_poll_quorum_rejected' panicked at 'VM error: RuntimeErr { msg: "Wasmer runtime error: InvokeError(TrapCode { code: Unreachable, srcloc: 194521 })", backtrace: Backtrace(()) }', /home/ethan/.cargo/registry/src/github.com-1ecc6299db9ec823/cosmwasm-vm-0.9.2/src/testing/calls.rs:43:5
...
16: cosmwasm_vm::testing::calls::handle
             at /home/ethan/.cargo/registry/src/github.com-1ecc6299db9ec823/cosmwasm-vm-0.9.2/src/testing/calls.rs:43
  17: integration::end_poll_quorum_rejected
             at tests/integration.rs:402
  18: integration::end_poll_quorum_rejected::{{closure}}
             at tests/integration.rs:348
```

This is a panic inside, at least I found it was in end_poll.
I then looked for "unwrap", which you should avoid, as you get no good error messages out - use `?` instead.

This lets the tests run without crashes:

```
        let staked_weight = 1234u128;
        // let staked_weight = deps
        //     .querier
        //     .query_balance(contract_address_human, &state.denom)
        //     .unwrap()
        //     .amount
        //     .u128();
```

**Comment by ethanfrey** [comment]
I cannot comment on code that is not modified here, but here are some changes that can improve it a bit, in `end_poll`

Replace

```rust
if (poll(&mut deps.storage).may_load(key.as_bytes())?).is_none() {
        return Err(StdError::generic_err("Poll does not exist"));
}

let mut a_poll = poll(&mut deps.storage).load(key.as_bytes()).unwrap();
```

with:

```rust
    let mut a_poll = poll(&mut deps.storage).load(key.as_bytes())?;
 ```

(does the same thing)

**Comment by ethanfrey** [comment]
After a bit more debugging, discovered the issue...

**Comment by ethanfrey** [comment]
```
        let quorum = ((tallied_weight / staked_weight) * 100) as u8;
```

And if staked_weight is 0? Rust triggers a panic.

2 solutions:
1. check that case and return proper error
2. set up the integration test with account balances

For (1) I would recommend to make use of [`Uint128 helpers`](https://github.com/CosmWasm/cosmwasm/blob/master/packages/std/src/math.rs#L283-L295). Edit: nevermind, this also panics on "divide by zero"...

Then just add a check, so:

```rust
        if staked_weight == 0 {
            return Err(StdError::generic_err("Cannot divide by zero - staked_weight"));
        }
        let quorum = ((tallied_weight / staked_weight) * 100) as u8;
```

Now, you get proper error messages

**Comment by ethanfrey** [comment]
The issue (which I will comment on the code) is that you didn't set a balance for the contract in the integration test:

In `tests.rs`:

```rust
    #[test]
    fn end_poll_quorum_rejected() {
        let mut deps = mock_dependencies(20, &coins(100, VOTING_TOKEN));
```

In `integration.rs`:

```rust
#[test]
fn end_poll_quorum_rejected() {
    let stake_amount = 100;
    let mut deps = mock_instance(WASM, &[]);
```

File: voting/tests/integration.rs [hunk 1]
```diff
@@ -17,3 +18,873 @@
...
+            log("rejected_reason", "Quorum not reached"),
+            log("passed", "false"),
+        ]
+    );
+}
+
+#[test]
+fn end_poll_quorum_rejected() {
+    let stake_amount = 100;
+    let mut deps = mock_instance(WASM, &[]);
```
Comment (ethanfrey) [inline voting/tests/integration.rs]: `let mut deps = mock_instance(WASM, &coins(stake_amount, VOTING_TOKEN));`

File: voting/tests/integration.rs [hunk 2]
```diff
@@ -17,3 +18,873 @@
...
+        ]
+    );
+}
+
+#[test]
+fn end_poll_nay_rejected() {
+    let voter1_stake = 100;
+    let voter2_stake = 1000;
+
+    let mut deps = mock_instance(WASM, &[]);
```
Comment (ethanfrey) [inline voting/tests/integration.rs]: `let mut deps = mock_instance(WASM, &coins(voter1_stake, VOTING_TOKEN));`

File: voting/tests/integration.rs [hunk 3]
```diff
@@ -17,3 +18,873 @@
...
+    }
+}
+
+#[test]
+fn happy_days_end_poll() {
+    const POLL_END_HEIGHT: u64 = 1000;
+    const POLL_ID: u64 = 1;
+    let stake_amount = 1000;
+
+    let mut deps = mock_instance(WASM, &[]);
```
Comment (ethanfrey) [inline voting/tests/integration.rs]: `let mut deps = mock_instance(WASM, &coins(stake_amount, VOTING_TOKEN));`

**Review by ethanfrey — COMMENTED** [review]
With these fixes, tests should pass (did locally for me).
Also fix up the 0 denominator check in `end_poll` as mentioned in the comments.

The big issue here is that panic in wasm code gives us no usable info, so make sure to always error and not panic. `?` instead of `.unwrap()` and check that math (`5/0` or `5u128 - 10u128` will panic)

**Comment by levackt** [comment]
Awesome digging there, thanks!

**Comment by levackt** [comment]
squashed these, closing to open the PR in a clean branch

**Pull request closed without merging**